testing-and-tdd/jasmine-testing.md

```diff
@@ -12,67 +12,176 @@ Here are links to lessons that should be completed before this lesson:
+5. Expect inside of asynchronous code is ignored, therefore passing. (false positive)
+- Solve this problem ( in jasmine and mocha ) with a parameter like `done`. Signaling to the test engine this is asynchronous code and it must wait. Mocha will also allow `return Promise`... inside the function which gives a similar signal to the test engine to wait for async code.
 
 ### Guided Practice
+*FIRST PRACTICE*
+1. Create a project using `npm init`.
+2. Install `jasmine` as `devDependencies`
+3. Create a function in a `myFunction.js` file. Do not complete the function. We will start with the function failing our test.
+4. Setup a test file that will test `myFunction.js` returns the expected value.
+5. Complete the function so that it returns a value and passes the test.
+
+*SECOND PRACTICE*
+1. Create a different project using `npm init`.
+2. Install `mocha`, and `chai` as `devDependencies`.
+4. Create a function in a `yourFunction.js` file. Same as above do not complete the function.
+5. With Mocha/Chai test that `yourFunction.js` returns a value, use chai `assert`.
+6. Complete the function so that it returns a value and passes the test.
 
-Have the apprentices work with you as you do something.
+### Independent Practice
 
+*Instructions:* Write your own test that validates an email address was passed to the function `userEmail(varString)`. The function is passed a string, so create this function and test it.
 
-### Independent Practice
+*Supplemental resources:*
+- JS testing with Jasmine blog post, [link](https://automationpanda.com/2018/01/26/javascript-testing-with-jasmine/#project_structure)
+- Chai and Mocha blog post, [link](https://codeburst.io/javascript-unit-testing-using-mocha-and-chai-1d97d9f18e71)
+  - Accompanied by this github repo of tests, [link](https://github.com/npatro/javascript-unit-testing-with-mocha)
+- Video series by "WebDevJourney"
+  - [Node JS - Authentication - Login/Logout Mocha Testing by Web Dev Journey](https://youtu.be/2q62EhBy0sE)
+  - github repo with mocha test included, [link](https://github.com/webdevjourneyWDJ/Todo-API).
 
-Class does this thing themselves with specific additional items.
+- Video by Dylan C. Israel [Unit Testing in JavaScript and Jasmine](https://www.youtube.com/watch?v=h2eWfvcAOTI) *starts at 12:00 minutes to show mocha test.*
 
 
 ### Challenge
 
-Apprentices can try to do this other thing.
+*Instructions:* Make your independent practice test run in the browser. The challenge is the environment must be modified for this test to run in the browser instead of the command line.
 
 
 ### Check for Understanding
 
-Some ideas: have apprentices summarize to each other, make a cheat sheet, take a quiz, do an assignment, or something else that helps assess their understanding.
+*Question:* What is Jasmine?
+*Answer to Q:* Jasmine is a behavior-driven development (BDD) for testing JavaScript code. Jasmine has no external dependencies and does not require a DOM.
```

Comment:
Maybe have this elsewhere in the curriculum, that way they're not looking right at it as they answer?

Reply:
I moved this to the top for when the students are first introduced to Jasmine. Plus I changed the question a little.

*Question:* What is Jasmine and how does it work in your project?